[1단계 - 콘솔 기반 로또 게임] 루나(오신정) 미션 제출합니다.

ShinjungOh opened the pull request:
# 🎰 콘솔 기반 로또 게임

안녕하세요 쵸파! 이번 미션의 리뷰이 루나라고 합니다. 콘솔 기반 로또 게임 미션을 완성해서 리뷰 요청 드립니다! 

기능 요구사항에 맞게 정상적으로 동작하는 코드를 작성했지만, 함수의 매개변수가 3개인 `retryUntilValid`와 같은 코드도 있고 아직 프로그래밍 요구 사항을 모두 준수하고 있지는 않습니다. 리팩토링 과정에서 프로그래밍 요구사항을 만족하도록 해보겠습니다.💪

## 학습 목표
이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.
- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트  
<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->
- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [ ] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?

## 리뷰 요청 & 논의하고 싶은 내용  
<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.  
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? --> 

1. TDD 테스트할때 이미 구현해야 하는 기능이 다 정해져있어서 중간에 기능을 바꾸는게 어려웠습니다. 
TDD에 익숙하지 않아서 그런 것인지, 처음에 설계를 잘못한 것인지, 테스트 코드를 도중에 바꿔도 될지 모르겠습니다.(TDD 너무 어렵네요😭) 
해결을 하지는 못했지만 가장 많이 고민했던 문제이기 때문에 이번 항목에 적습니다.

2. 상수의 구조를 어떻게 작성할지 고민했습니다. 
error의 경우, MESSAGE에 입출력 메시지와 함께 넣을지 ERROR를 따로 분리할지 고민했는데, 결국 메시지라는 관심사로 묶이기 때문에 한 파일에 두기로 결정했습니다. 또한 각각을 `export const ~`로 상수화할지, 객체를 생성해서 객체 내부에 다 넣을지 고민했는데 매번 상수를 가져다 사용할 때 매번 import를 해줘야 하는 것이 번거로울 것 같아 후자의 방법을 택했습니다.


### 2) 이번 리뷰를 통해 논의하고 싶은 부분  
<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->


1. 파라미터의 순서 
함수의 파라미터는 중요한 순서대로 넘겨주는 것으로 알고 있습니다. 
`bonusNumberValidator` 에서 인자 순서를 `bonusNumber`, `winningNumbers`로 주고 있는데, 이 순서가 괜찮은지 궁금합니다. 프로그램의 맥락상으로는 사용자에게 로또 당첨 번호를 입력하고 보너스 번호를 받고 있습니다. 그렇지만 `bonusNumberValidator` 파일은 보너스 번호를 검증하는 역할을 하고 있기 때문에, 더 중요한 인자는 보너스 번호이지 않을까 하는 의문이 들었습니다. 때문에 지금의 순서가 맞나?하는 고민이 생겼는데 어떻게 생각하시는지 궁금합니다!

```js
export const bonusNumberValidator = (bonusNumber, winningNumbers) => {
  // ... 
};
```

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계
- [x] TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [x] 도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [x] 객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [ ] 불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

</details>

mlnwns committed:
docs: coauthor 템플릿 작성

Co-authored-by: ShinjungOh <ShinjungOh@users.noreply.github.com>

mlnwns committed:
​docs: 기능 명세서 작성

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
chore​: 폴더 구조 설계

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
​feat: 입출력, 에러 메시지 및 로또 세팅 상수화

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
feat: 사용자 정의 오류 클래스 추가​

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
feat: 비동기 입력 처리 기능 추가

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
feat: 출력 기능 추가

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
​feat: 구입금액 입력 기능 추가

- LottoController 생성
Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
​feat: 재입력 함수 기능 구현

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
​feat: 로또 구입 금액 입력 로직 및 검증 구현

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
feat: 구입금액 정수로 파싱​

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
feat: 구입 금액에 따른 로또 갯수별 발행 구현

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
​feat: 구입 금액에 해당하는 만큼 로또 출력

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
​refactor: 불필요한 parse 문법 제거

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
feat: 당첨 번호 입력 및 유효성 검사 구현​

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
​refactor: 숫자 입력 검증 내 에러 메시지 변경
- 소수 입력 불가 메시지 추가
- 구입금액, 당첨 번호 에러 메시지 변경 적용

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
​feat: 보너스 번호 입력 기능 구현

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
​feat: 로또 번호와 당첨 번호를 비교해서 당첨 개수 계산

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
​feat: 당첨 내역 및 수익률 출력

Co-authored-by: ShinjungOh <97431021+ShinjungOh@users.noreply.github.com>

mlnwns committed:
feat: 재시작/종료 여부 입력에 따라 프로그램 재시작 결정 및 실행

Co-authored-by: ShinjungOh <ShinjungOh@users.noreply.github.com>

bassyu commented on src/domain/LottoGenerator.js:
해당 로직을 domain으로 분류해주신 것 좋네요 👍 
다만, class로 작성해주신 이유가 궁금합니다!

ShinjungOh replied:
사실 클래스와 함수형 중 어떤 방식으로 사용할지 큰 고민을 하지 않았던 것 같아요🥲 기존에 프리코스 및 이전 미션에서 클래스로 쭉 사용해와서 이번에도 그대로 사용하게 되었습니다.

중요한 포인트는 제가 **클래스와 함수형이 각각 언제 사용되어야 좋을까**에 대한 지식이 부족해서 그렇다고 생각되어요.

이번 로또 미션에서 해당 부분을 고민하지 않고 작성한 부분이 많은데요, 클래스와 함수형의 차이를 학습하고 하나씩 개선점을 찾아서 리팩토링 해보겠습니다!

bassyu commented on src/utils/utils.js:
utils 함수이니, 부수효과를 제거해보셨으면 좋겠습니다!
혹시 왜 제거하려는지 아시나요?

ShinjungOh replied:
유틸함수는 재사용이 용이해야 하므로, 순수함수로 만들어 부수효과를 제거하면 좋겠다는 의미 같네요! 
지금 제가 작성한 방식은 원본 배열을 변경하는 방식이기 때문에 새로운 배열을 반환하는 형식으로 고쳐봐야겠어요.
제가 이해한 바가 맞을까요?

bassyu replied:
넵! 맞습니다! 👍 👍

bassyu commented on src/utils/retryUntilValid.js:
지금은 utils 함수이긴 하지만, 생각보다 역할이 많습니다!
output view도 알고 있네요!

정말 범용적으로 쓰려는 목적이 아니라면, 쿨하게 util 대신 View나 Controller 코드로 생각하는 건 어떤가요?
중복 제거의 의미도 있으니까요~

ShinjungOh replied:
View(Input)로 `retryUntilValid`의 코드를 옮기기로 했어요. 

사용자의 입력값을 받아 파싱하고, 유효성을 검사하고, 유효성 검사에서 에러가 발생하면 에러 메시지를 출력하는 기능이 `View(Input)`에 있어야 관심사 분리가 잘 된 것이라고 판단했어요.

bassyu commented on src/domain/LottoGenerator.js:
```diff
@@ -0,0 +1,28 @@
+  static createLottoNumbers() {
+    const numbers = new Set();
+
+    while (numbers.size < 6) {
+      const randomNumber = Math.floor(Math.random() * 45) + 1;
```
상수화 포인트가 많네요~!

ShinjungOh replied:
개발하다가 놓친 포인트네요😅 상수화 해놓겠습니다!

bassyu commented on src/domain/LottoMatcher.js:
생성자를 통해 딱 1번의 인풋을 받은 후, 이후엔 메서드로 반환만 하고있네요!
그래서 순수한 함수와 비슷하지 않나? 클래스여야 할까? 라는 생각이 들었어요.

ShinjungOh replied:
리뷰 보고 고민해봤는데 클래스여야만 할 이유가 없는 것 같아요! 
오히려 **함수형**으로 작성하는게 낫겠다는 생각이 들었는데요,  
제가 정리한 이유는 다음과 같아요.

1. 클래스의 상태 변화가 없고,
2. 인풋에 따라 항상 같은 값을 출력하는 순수함수이며,
3. 인스턴스를 생성할 필요가 없다.

앞서 `LottoGenerator.js`에서도 마찬가지의 이유로 함수형으로 리팩토링 했는데 계속 하다보니 감이 잡히는 것 같습니다🚀

bassyu replied:
상태 변화가 없다는 점이 포인트였는데 이해해주셨네요 👍

bassyu commented on src/validators/purchaseAmountValidator.js:
다른 크루에게도 드렸던 코멘트라 [링크](https://github.com/woowacourse/javascript-racingcar/pull/338/files/b09f6e02313896720938b51b564a75e58f070d3f#r1954575955)로 대체할게요!

이후 [유연한 Validate 함수 예제](https://github.com/woowacourse/javascript-racingcar/pull/410)가 있어서, 고민해보신 다음에 참고해봐도 좋을 것 같아요!

ShinjungOh replied:
> '꼭 Validator와 input 함수가 에러로 소통해야 할까?'라는 생각이 들었습니다!
잘못된 입력을 알려주는 것도 View 영역이라면, 판단은 Validator가 하고, 어떤 문자열을 보내줄지는 input 함수에서 책임지는 방법은 어떤가요?

이 말의 의미를 제대로 이해하지 못하겠어요😭 
링크로 주신 함수 예제 또한 저한테는 어렵게 느껴지네요..

'Validator는 true 또는 false로 유효성 검사가 통과했는지만 알려주는 것이 어떨까?' 라는 의미이실까요?🤔

bassyu replied:
넵 그런 의미 맞습니다!

bassyu commented on src/domain/LottoGenerator.test.js:
```diff
@@ -0,0 +1,48 @@
+import { sortNumber } from "../utils/utils.js";
+
+describe("LottoGenerator 테스트", () => {
+  test("입력된 금액에 따라 올바른 개수의 로또를 생성해야 한다", () => {
```
입력된 금액이 정확히 얼마면, 몇개를 생성해야 하는지 자세히 알려줄 수 있습니다!
test.each 함수를 활용해서 다양한 케이스를 테스트 해볼 수도 있겠네요!

ShinjungOh replied:
변수명만큼 테스트 이름 짓는것도 어렵더라구요. 제출하고 나니 스스로 아쉬운 부분이 보이기도 했구요. 
리뷰 반영하면서 테스트 이름을 신경써서 고쳐보겠습니다! 
`test.each`로 다양한 케이스 테스트 해보는 것도 반영해볼게요~!

bassyu reviewed (changes requested):
안녕하세요 루나!!
로또 미션 잘부탁드려요! 🙇 

## 코드

전반적으로 잘 작성해주셨네요~! 👍 
특히 도메인 단위를 잘 나눠주셨고, 덕분에 테스트 코드도 잘 작성된 것 같습니다!
검증 테스트도 잘해주셨네요! 💯

다만, 클래스 사용에 관련해서는 코멘트 드렸습니다~!

## QnA

> TDD에 익숙하지 않아서 그런 것인지, 처음에 설계를 잘못한 것인지, 테스트 코드를 도중에 바꿔도 될지 모르겠습니다.(TDD 너무 어렵네요😭)

팁을 드리자면, 함수의 단위를 우선 최대한 작게 나누어보세요! 충분히 작은 단위라면, 테스트코드를 수정하는 일이 줄어들 것 같습니다~!

> error의 경우, MESSAGE에 입출력 메시지와 함께 넣을지 ERROR를 따로 분리할지 고민했는데, 결국 메시지라는 관심사로 묶이기 때문에 한 파일에 두기로 결정했습니다.

좋은 것 같습니다! 추가로 그냥 메세지보다 더 확실하게 'View와 관련있는 메세지 상수'로 범위를 좁혀서 네이밍하거나 폴더를 옮기는 것도 좋을 것 같아요.

> bonusNumberValidator 에서 인자 순서를 bonusNumber, winningNumbers로 주고 있는데, 이 순서가 괜찮은지 궁금합니다.

지금도 괜찮은 것 같아요!
보너스라는 의미 자체가 winningNumbers가 준비된 상태여야 하니, 인자를 2개 받는 부분도 좋은 것 같습니다.

ShinjungOh committed:
refactor: sortNumber 유틸 함수의 부수효과 제거

ShinjungOh committed:
refactor: LottoGenerator의 매직 넘버를 상수화

ShinjungOh committed:
test: test.each를 이용해 구입 금액에 따른 로또 갯수 검증

- restartValidator 테스트 이름을 상황에 맞게 구체적으로 개선
- 테스트 이름에 마침표가 빠진 경우를 찾아 마침표 추가

ShinjungOh committed:
refactor: LottoGenerator 함수형으로 변경

- LottoGenerator의 함수형 리팩토링에 따른 변화를 테스트 코드에도 적용
- LottoGenerator 테스트 코드의 매직 넘버를 상수화

ShinjungOh committed:
refactor: LottoMatcher 함수형으로 변경

- LottoMatcher의 함수형 리팩토링에 따른 변화를 테스트 코드에도 적용

ShinjungOh committed:
refactor: Input 클래스를 함수형으로 변경

ShinjungOh committed:
refactor: Output 클래스를 함수형으로 변경

ShinjungOh committed:
refactor: retryUntilValid 로직을 Input View로 이동

ShinjungOh committed:
feat: 최대 구입 가능 금액을 10만 원으로 제한

- purchaseAmountValidator 로직과 테스트에 반영
- 10만원을 상수로 분리

ShinjungOh committed:
refactor: line length가 100 이상인 코드의 줄바꿈 처리

ShinjungOh commented:
# 🎰 콘솔 기반 로또 게임 - 리팩토링

안녕하세요 쵸파! 리팩토링 마치고 리뷰요청 드립니다. 🙇‍♀️
댓글로 남겨주신 리뷰와 다른 크루들의 코드리뷰를 바탕으로 리팩토링을 진행했습니다. 
주요 변경 사항과, 리팩토링을 진행하며 궁금했던 점 남겨둘게요! 

<br>

## 🎬 실행 영상

https://github.com/user-attachments/assets/c423b878-d406-4367-8935-b85747d09238

<br>

## ⚙️ 주요 변경 사항

### 1. 클래스를 함수형으로 변경

[96f33e1](https://github.com/woowacourse/javascript-lotto/pull/350/commits/96f33e119e6ee0000a6e3e21963adc1859114f5f)
[ff894ce](https://github.com/woowacourse/javascript-lotto/pull/350/commits/ff894cee42049393028c9064ece6591f645fa5e2)
[1fed511](https://github.com/woowacourse/javascript-lotto/pull/350/commits/1fed5117cf44ed1cc09552e466d37d1b290484b3)
[d7c33ab](https://github.com/woowacourse/javascript-lotto/pull/350/commits/d7c33abc302909b7bf15a8052d108bb04bd6a606)

해당하는 리팩토링은 `LottoGenerator`, `LottoMatcher`, `Input`, `Output` 클래스에서 진행되었습니다. 
리팩토링 하면서 클래스와 함수형의 차이를 정리해보았어요.

1. 클래스 
클래스는 상태(프라이빗 필드)가 존재하고, 상태가 변경될 때, 인스턴스를 생성할 필요가 있을 때 사용합니다. 
같은 관심사를 가진 메소드가 얽혀있어서 클래스라는 틀로 동일한 객체를 여러개 생성해줄 때는 필요하지만, 
그 외에는 범용적으로 사용하기 힘들어집니다. 결과적으로 재사용성이 떨어지고 유지보수하기 힘든 코드가 됩니다.  

2. 함수형
함수는 상태를 가지고 있지 않으며, 한 가지 기능만 수행합니다. 이를 모듈처럼 사용해서 재사용할 수 있기 때문에 유지보수에도 편리해집니다.

리뷰로 달아주신 `LottoGenerator`, `LottoMatcher` 외에도 View를 함수형으로 개선할 수 있을 것 같아서 리팩토링을 진행했어요. 
상태가 없고, 인스턴스를 생성하지 않는 클래스는 함수형으로 바꿔도 될 것 같다는 판단을 내렸습니다!

### 2. sortNumber 유틸 함수의 부수효과 제거

[17e940e](https://github.com/woowacourse/javascript-lotto/pull/350/commits/17e940e341fbcd4245986fa40b328d82940bd05b)

원본 배열을 변경하는 방식으로 작성된 로직을 새로운 배열을 반환하는 형식으로 개선했습니다. 
유틸 함수이기 때문에 순수함수로 만들어 부수효과를 제거했습니다.
원본 배열을 직접 변경하지 않기 때문에 안전하고, 유지보수하기 쉬워졌습니다. 

### 3. retryUntilValid 로직을 Input View로 이동

[ae95399](https://github.com/woowacourse/javascript-lotto/pull/350/commits/ae95399b099fc1ecec94fe27b0817a8d1399c4c0)

1. 사용자의 입력값을 받아 파싱한다
2. 입력값의 유효성을 검사한다. 
4. 유효성 검사에서 에러가 발생하면 에러 메시지를 출력한다

이 기능이 하나의 함수에 묶여있고, 제가 생각하기에는 Input의 관심사를 가지고 있다고 판단했어요.
유틸함수라기로 보기에는 View와 관련되어 있어서 `View(Input)`으로 이동시켰습니다.


### 4. 매직 넘버의 상수화

[f2f3a3e](https://github.com/woowacourse/javascript-lotto/pull/350/commits/f2f3a3e988072eea77cc9f4f9d60ff4e324afe1f)  
[fe05150](https://github.com/woowacourse/javascript-lotto/pull/350/commits/fe05150a0639b85f6432a87de7beea2921c16380)

상수화 할 수 있는 포인트를 상수화 해두었어요.

### 5.  로또의 최대 구입 가능 금액을 10만 원으로 제한 

[fe05150](https://github.com/woowacourse/javascript-lotto/pull/350/commits/fe05150a0639b85f6432a87de7beea2921c16380)

한국에서 로또는 1인당 1회 최대 10만원까지 구매할 수 있다고 하는 뉴스가 있어서, 저도 10만원으로 제한을 두었습니다😄
[[팩트체크] 로또복권은 현금으로만 구입할 수 있다?](https://www.yna.co.kr/view/AKR20250109060300518)

원래는 최대 구매 가능 금액에 제한을 두지 않았었는데 개발할 때 이 부분을 놓쳐서 기능 추가했어요!

### 6. test.each를 이용해 구입 금액에 따른 로또 갯수 검증

[a2c8020](https://github.com/woowacourse/javascript-lotto/pull/350/commits/a2c80203f6cc754bcadc6bd2002ca984e9bbc45d)

`test.each`로 "금액 %i원이 입력되면 로또 %i개를 생성한다."는 테스트를 진행했어요. 
이번 미션의 테스트에서 `%i, %s`를 사용했는데 이 문자를 뭐라고 하는지 몰랐다가 test.each의 `placeholder`라고 말한다는걸 알게 되었어요. 
해당 부분을 앞으로도 자주 사용하게 될 것 같아서 TIL로 간단하게 정리해두었어요.

[TIL: Jest test.each의 placeholder](https://github.com/ShinjungOh/TIL/blob/main/Jest%20-%20test.each%EC%9D%98%20placeholder.md)

그 외 테스트 이름이 명확하지 않거나, 마침표를 찍지 않은 테스트명에 마침표를 추가하는 등의 리팩토링을 진행했습니다.

<br/>

## 🔎 리뷰어에게 궁금한 것

1.  **함수형 모듈**로 변경한 코드와 **객체**로 표현한 것 중 어떤 것이 나은가?
`Ouput`을 예시로 들자면 다음과 같습니다.

```js
// 기존 코드

export default class Output {
  static print(message) {
    console.log(message);
  }

  static lottoResult() {
    this.print(MESSAGES.output.result);
    this.print(MESSAGES.output.divider);
  }

  static printRankResult(count, description) {
    this.print(`${description} - ${count}개`);
  }
}
```

위의 기존 코드를 개선한다고 했을 때, 아래의 두 방법 중 어느 것을 선호하시는지 궁금합니다. 

```js
// 리팩토링한 코드 - 함수형으로 변경

export const outputPrint = (message) => {
  console.log(message);
};

export const lottoResult = () => {
  outputPrint(MESSAGES.output.result);
  outputPrint(MESSAGES.output.divider);
};

export const printRankResult = (count, description) => {
  outputPrint(`${description} - ${count}개`);
};
```

```js
// 객체로 표현할 경우 

export const output = {
  outputPrint(message) {
    console.log(message);
  },

  lottoResult() {
    this.outputPrint(MESSAGES.output.result);
    this.outputPrint(MESSAGES.output.divider);
  },

  printRankResult(count, description) {
    this.outputPrint(`${description} - ${count}개`);
  },
};
```

저는 필요한 함수만 모듈화해서 사용할 수 있도록 **함수형** 방식을 채택했습니다. 
그렇지만 함수형으로 바꾸고 나니 파일명을 어떻게 바꿔야 할지 또는 그냥 놔둬야 할지 고민이 되었어요. (현재는 파일명이 그대로인 상태) 
함수가 어느 관심사에 속하는지 보여주기 위해서는 객체형을 사용해서 `Output.lottoResult`로 호출해야 보기 편하다는 생각도 들었습니다. 

앞서 `LottoGenerator`, `LottoMatcher`도 함수형으로 변경한 관계로 View 역시 함수형으로 바꾸는 방식을 선택했는데, 어떻게 생각하시는지 궁금합니다.

2. Validator에서 에러 메시지를 출력하는 것이 좋지 않은 방식인가?

리뷰로 달아주신 https://github.com/woowacourse/javascript-lotto/pull/350#discussion_r1963632045 부분이 이해가 잘 되지 않았습니다. 
'Validator는 true 또는 false로 유효성 검사가 통과했는지만 알려주는 것이 어떨까?' 라고 제안하신 걸로 받아들였는데요,
저는 지금 구현되어있는 '현재 각 유효성 검증 상황에 맞는 에러 메시지 출력' 방식이 직관적이고 테스트할 때도 명확하다고 생각해서 이렇게 사용해 왔어요. 
그래서 리뷰가 어렵게 느껴졌는데, 리뷰 주신 내용을 조금 더 설명해 주신다면 비교해서 고민해 볼 수 있을 것 같아요!


<br/>

bassyu reviewed (approved):
안녕하세요 루나!!
코멘트 많이 반영해 주시면서 고민도 많이 하시고, 다양한 시도도 해보셨군요! 💯 💯

# 코드

함수와 클래스를 써보시면서 '인스턴스가 필요한지'를 기준으로 잡으셨군요! 
좋은 기준 중 하나네요 💯 

> 5. 로또의 최대 구입 가능 금액을 10만 원으로 제한

좋습니다! 이렇게 코드 뿐만 아니라 서비스에도 애매한 부분은 충분히 생길 수 있는데요!
애매한 부분 상상하고 고려하면서 대응하는 자세가 좋다고 생각합니다~! 👍 

## QnA

> 위의 기존 코드를 개선한다고 했을 때, 아래의 두 방법 중 어느 것을 선호하시는지 궁금합니다.

저는 기본적으로 1번, 개수가 좀 많고 공통점이 있으면 2번을 사용합니다!

> 앞서 LottoGenerator, LottoMatcher도 함수형으로 변경한 관계로 View 역시 함수형으로 바꾸는 방식을 선택했는데, 어떻게 생각하시는지 궁금합니다.

루나 기준에 일관성을 더하는 느낌이어서 좋습니다~!

> Validator에서 에러 메시지를 출력하는 것이 좋지 않은 방식인가?

다시 질문 주셔서 감사합니다 ㅎㅎ
지금 구조에선 이대로도 괜찮습니다! 확장성을 고려해서 리뷰 드렸던 내용이에요~!

예를 들어서, 번호 입력 중 3개의 검증을 한번에 실패(중복, 개수, 범위)했을 때, "중복도 있고, 개수도 안맞고, 범위도 넘었어요"를 출력해야 한다면..?
'검증'과 '검증 결과를 활용' 코드가 나뉘어져 있다면 더 유연하게 대처할 수 있습니다~

---

고생 많이 하셨어요!! 1단계는 이만 머지하겠습니다! 👍 
2단계도 파이팅입니다~! 💪 💪
